Zend/zend_attributes.h

```diff
@@ -37,8 +35,59 @@ ZEND_API zend_attribute *zend_get_parameter_attribute_str(HashTable *attributes,
 ZEND_API void zend_compiler_attribute_register(zend_class_entry *ce, zend_attributes_internal_validator validator);
 ZEND_API zend_attributes_internal_validator zend_attribute_get_validator(zend_string *lcname);
 
-void zend_register_attribute_ce(void);
+ZEND_API zend_attribute *zend_add_attribute(HashTable **attributes, zend_bool persistent, uint32_t offset, zend_string *name, uint32_t argc);
 
 END_EXTERN_C()
 
+static zend_always_inline zend_attribute *zend_add_class_attribute(zend_class_entry *ce, zend_string *name, uint32_t argc)
+{
+	return zend_add_attribute(&ce->attributes, ce->type != ZEND_USER_CLASS, 0, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_function_attribute(zend_function *func, zend_string *name, uint32_t argc)
+{
+	return zend_add_attribute(&func->common.attributes, func->common.type != ZEND_USER_FUNCTION, 0, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_parameter_attribute(zend_function *func, uint32_t offset, zend_string *name, uint32_t argc)
+{
+	return zend_add_attribute(&func->common.attributes, func->common.type != ZEND_USER_FUNCTION, offset + 1, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_property_attribute(zend_class_entry *ce, zend_property_info *info, zend_string *name, uint32_t argc)
+{
+	return zend_add_attribute(&info->attributes, ce->type != ZEND_USER_CLASS, 0, name, argc);
+}
+
+static zend_always_inline zend_attribute *zend_add_property_attribute_str(zend_class_entry *ce, const char *key, size_t len, zend_string *name, uint32_t argc)
```

Comment:
I'd accept a zend_string here instead, as the more flexible option.

Reply:
I did not use zend_string because I thought the most common use case would be like this:
```c
zend_add_property_attribute_str(my_ce, ZEND_STRL("prop1"), attr_ce->name, 0);
zend_add_property_attribute_str(my_ce, ZEND_STRL("prop2"), attr_ce->name, 0);
```
Allocating a lot of strings seemed like more work for API users. But if we change the registration functions to return pointers to structures this would give easy access to the name as zend_string which would be even better.

Reply:
If we return pointers, there's probably no need for the _str APIs at all.